### ISSUE TITLE
pip3 install fails on Ubuntu 20.04?

### ISSUE THREAD

**Issue by niess** [tudo-astroparticlephysics/PROPOSAL#193]
Hello,

first let me thank you for this great work and for making it available to the community :)

I had been using version 6.1.5 of PROPOSAL on Debian 9 without issues as far as I was concerned. Recently, I wanted to upgrade to the latest PyPI version of PROPOSAL, i.e. 7.0.5. But without success so far :(

Since Debian 9 seems too old w.r.t. several deps of PROPOSAL I ended up trying Ubuntu 20.04 first through docker and then through VirtualBox. In both cases I got the following issue. Installing with pip3 as:

```
pip3 install proposal
```
runs smoothly. However, importing the `proposal` Python package crashes as:

```Python
ImportError: $HOME/.local/lib/python3.8/site-packages/proposal.cpython-38-x86_64-linux-gnu.so: undefined symbol: _ZN13cubic_splines12CubicSplinesIdEC1ERKNS1_10DefinitionENSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESA_
```
Peeking at `proposal.cpython-38-x86_64-linux-gnu.so` with objdump indicates:
```
Dynamic Section:
  NEEDED               libpthread.so.0
  NEEDED               libstdc++.so.6
  NEEDED               libm.so.6
  NEEDED               libgcc_s.so.1
  NEEDED               libc.so.6
  NEEDED               ld-linux-x86-64.so.2
  NEEDED               libpython3.8.so.1.0
  SONAME               proposal.cpython-38-x86_64-linux-gnu.so
  RUNPATH              $ORIGIN
```
which does not exhibit any missing deps.

Note that with VirtualBox the Linux kernel is 5.11. Thus, this does not look like issue #186 . Python version is 3.8.10 and gcc version is 9.4 in case it could help?

Am I doing something wrong? What OS would you recommend to use?

**Comment by Jean1995** [comment]
Hello! 

This sounds very much like a problem with the package manager conan. We have already experienced these problems on severral Linux systems before.

Could you try to run `conan profile update settings.compiler.libcxx=libstdc++11 default`, and then try to install proposal using `pip3 install proposal` again?

If this does not already help, there might be some cached files left that need to be rebuild. In this case, try to follow the steps that I suggest in issue #180.

I hope that these steps can already solve your issue!

**Comment by niess** [comment]
@Jean1995 Thanks. 

This seems to solve the issue :) Now I can import proposal and its seems to run smoothly.

It might be worth to add this info to the installation instructions. For the record here is what I did.

```
pip3 uninstall proposal
pip3 install conan
# conan profile new default --detect
conan profile update settings.compiler.libcxx=libstdc++11 default
pip3 install proposal
```

The commented line is needed in order to create the default profile if it does not already exist. I needed it when building an image with Docker or Singularity.

You might be interested in the following as well. When running proposal from a Docker or Singularity image the host kernel is used, i.e. 4.9 in my case. Yet, it seems to work properly using the recipe above. Thus, this might help with issue #186 as well ?

**Comment by maxnoe** [comment]
Let me try to understand what is happening here.

Is the problem that cubic interpolation was compiled linking to `libstdc++` (gcc) and then proposal was linked against `libc++` (clang)? That should not happen and that would hint to a problem with the build systems. The solution cannot be to manually force proposal to link the gnu c++ std lib. 

Why does conan / cmake link cubic interpolation to libstc++ and then proposal to libc++? There has to be an error in the build / package configuration.

**Comment by Jean1995** [comment]
> Is the problem that cubic interpolation was compiled linking to `libstdc++` (gcc) and then proposal was linked against `libc++` (clang)? That should not happen and that would hint to a problem with the build systems. The solution cannot be to manually force proposal to link the gnu c++ std lib.
> 
> Why does conan / cmake link cubic interpolation to libstc++ and then proposal to libc++? There has to be an error in the build / package configuration.

From what I understand, conan provides different options to choose the library ABI that is used for the binaries. Per default, this is set to `libstdc++`, but can be set to, for example, `libstdc++11`: [https://docs.conan.io/en/latest/howtos/manage_gcc_abi.html](https://docs.conan.io/en/latest/howtos/manage_gcc_abi.html)

[Here](https://gcc.gnu.org/onlinedocs/libstdc++/manual/using_dual_abi.html), it is said that such a linker error hints that the libraries were compiled with different compiler versions or different ABI settings.

However, if I keep the ABI setting at `libstdc++`, I am not sure whether this error is expected. Shoudn't this mean that all libraries and dependencies are compiled with the same ABI? Or is it a problem when you locally use a gcc version that uses `libstdc++11` per default, and then try to use binaries that were compiled with `libstdc++`?

I think it is very important to solve this issue asap, since the (vast?) majority of users will probably run into this problem, especially when installing proposal using pip.

**Comment by Jean1995** [comment]
> However, if I keep the ABI setting at `libstdc++`, I am not sure whether this error is expected. Shoudn't this mean that all libraries and dependencies are compiled with the same ABI? Or is it a problem when you locally use a gcc version that uses `libstdc++11` per default, and then try to use binaries that were compiled with `libstdc++`?

I think the problem basically is that, without any extra specification, conan is using the binaries for the dependencies build with libstdc++, while PROPOSAL is built locally. But the local compiler uses libstdc++11 if the gcc version is >=5. This means that PROPOSAL is built using libstdc++11, the dependencies (including cubicinterpolation) are built using libstdc++, and we get our linker error.

**Comment by Jean1995** [comment]
This should be fixed now with the release of version 7.0.7 where PR #195 is included!
Thank you very much for the report!
